Decision making & contributing (#81)
bids-standard/bids-matlab: CONTRIBUTING.md

  - [Writing in markdown](#writing-in-markdown)
  - [Matlab code style guide and quality](#matlab-code-style-guide-and-quality)
- [Making a change with a pull request](#making-a-change-with-a-pull-request)
- [Example pull request](#example-pull-request)
- [Commenting on a pull request](#commenting-on-a-pull-request)
- [Accepting suggestion from a review](#accepting-suggestion-from-a-review)
- [Recognizing contributions](#recognizing-contributions)

<!--
TODO: sections to add
- testing
- examples
  - setting up an octave jupyter notebook
- binder
- continuous integration
  - travis
  - github actions
-->

## Joining the community

BIDS - the [Brain Imaging Data Structure](https://bids.neuroimaging.io/) - is a
growing community of neuroimaging enthusiasts, and we want to make our resources
accessible to and engaging for as many researchers as possible.

Most of our discussions take place here in
[GitHub issues](#understanding-issues).

To keep on top of new posts, please see this guide for setting your
[topic notifications](https://meta.discourse.org/t/discourse-new-user-guide/96331#heading--topic-notifications).

As a reminder, we expect that all contributions adhere to our

> [Code of Conduct](CODE_OF_CONDUCT.md).

Author: ```suggestion
[Code of Conduct](./CODE_OF_CONDUCT.md).
```

# Contributing to the BIDS-matlab

> **Welcome to the BIDS-matlab repository!**

Reviewer: Is the canonical styling "BIDS-matlab" or "BIDS-MATLAB"? I've seen it as "BIDS-MATLAB" other places in this repo, and MathWorks styles it "MATLAB".

Author: I honestly have no idea and no strong opinion about this.

We can go with `BIDS-MATLAB`.

Author: Will do a search and replace on the whole PR.

## Table of contents

Been here before? 
Already know what you're looking for in this guide? 
Jump to the following sections:

-   [Joining the community](#joining-the-community)
-   [Contributing through GitHub](#contributing-through-github)
-   [Understanding issues](#understanding-issues)
-   [Making a change with a pull request](#making-a-change-with-a-pull-request)
-   [Example pull request](#example-pull-request)
-   [Commenting on a pull request](#commenting-on-a-pull-request)
-   [Accepting suggestion from a review](#accepting-suggestion-from-a-review)
-   [Recognizing contributions](#recognizing-contributions)

<!-- 
TODO: sections to add
- testing
- examples
  - setting up an octave jupyter notebook
- binder
- continuous integration 
-->

## Joining the community

BIDS - the [Brain Imaging Data Structure](https://bids.neuroimaging.io/) - is a
growing community of neuroimaging enthusiasts, and we want to make our resources
accessible to and engaging for as many researchers as possible.

Most of our discussions take place here in
[GitHub issues](#understanding-issues). 

To keep on top of new posts, please see this guide for setting your

> [topic notifications](https://meta.discourse.org/t/discourse-new-user-guide/96331#heading--topic-notifications).

Author: I am tempted to keep the link to Neurostars to encourage people to post there. The mattermost channel is a horrible way to search for past problems unless you start pinning all the important messages and this quickly becomes high maintainance heavy :-).

> [topic notifications](https://meta.discourse.org/t/discourse-new-user-guide/96331#heading--topic-notifications).

Reviewer: Agreed about Mattermost; it makes for a lousy forum or reference doc.

As a reminder, we expect that all contributions adhere to our

> [Code of Conduct](CODE_OF_CONDUCT.md).

Reviewer: Broken link. Should this PR be rebased on `main` to pick up the new CoC file?

Author: It should be fine now.

## Contributing through GitHub

[Git](https://git-scm.com/) is a really useful tool for version control.

> [GitHub](https://github.com/) sits on top of git and supports collaborative and

Reviewer: ```suggestion
[GitHub](https://github.com/) sits on top of Git and supports collaborative and
```

This may be pedantic, but the source control system is "Git" with a capital "G"; `git` with a lower-case "g" is the command-line tool for it.

Author: Thanks. I keep making this mistake. And by all means "be pedantic". :wink:

distributed working.

> We know that it can be daunting to start using git and GitHub if you haven't

Reviewer: ```suggestion
We know that it can be daunting to start using Git and GitHub if you haven't
```

on the command line, please [download and install](https://nodejs.org/en/download/) 
NodeJS.

#### 2. Install Remark-CLI and our style guide

Remark-CLI can be installed via [npm](https://www.npmjs.com/), which is part of
the NodeJS distribution.

To install the packages we use for our style guide, the following command will
work on most command lines:

```
npm install `cat npm-requirements.txt`
```

The equivalent command on PowerShell is:

```
npm install @(cat npm-requirements.txt)
```

#### 3. Fix the flagged document

Please go to the directory where the flagged file is and run remark like this:

```
remark flagged_file.md -o flagged_file_fixed.md
```

Please confirm this has fixed the file. To do this, please run this:

```
remark flagged_file_fixed.md --frail
```

This command will indicate whether this file now conforms to the style guide. 
If it passes, replace `flagged_file.md` with the contents of
`flagged_file_fixed.md`, add and commit the change:

```
mv flagged_file_fixed.md flagged_file.md
git add flagged_file.md
git commit -m 'STY: Fixed Markdown style' 
```-->

## Making a change with a pull request

> We appreciate all contributions to the BIDS Specification. **THANK YOU** for

Reviewer: ```suggestion
We appreciate all contributions to BIDS-matlab. **THANK YOU** for
```

helping us build this useful resource.

#### 1. Comment on an existing issue or open a new issue referencing your addition

This allows other members of the BIDS-matlab team to confirm that you
aren't overlapping with work that's currently underway and that everyone is on
the same page with the goal of the work you're going to carry out.

#### 2. [Fork](https://help.github.com/articles/fork-a-repo/) [this repository](https://github.com/bids-standard/bids-matlab) to your profile

> This is now your own unique copy of the BIDS matlab. 

Reviewer: ```suggestion
This is now your own unique copy of BIDS-matlab. 
```

Changes here won't affect anyone else's work, so it's a safe space to explore edits to the
code!

Make sure to
[keep your fork up to date](https://help.github.com/articles/syncing-a-fork/)

> with the master repository, otherwise you can end up with lots of dreaded

Reviewer: Should this be "parent" or "main" repository, instead of "master", since the term "master" is now out of favor on GitHub?

Author: Yes let's overthrow `master` especially here because it could lead to confusion with the `master` branch of some repo. I think we will "soon" rename the `master` branch into `main` to follow the move.

I think that `upstream` is one of those terms often used but that is VERY obscure when you start Git and Github. So I would say `parent` sounds better.

> with the master repository, otherwise you can end up with lots of dreaded

Author: ```suggestion
with the parent repository, otherwise you can end up with lots of dreaded
```

[merge conflicts](https://help.github.com/articles/about-merge-conflicts/).

#### 3. Make the changes you've discussed

Try to keep the changes focused. 
If you submit a large amount of work in all in one go it will be much more work 
for whomever is reviewing your pull request.
Please detail the changes you are attempting to make.

#### 4. Submit a [pull request](https://help.github.com/articles/about-pull-requests/)

Please keep the title of your pull request short but informative. 

<!-- It will appear in the [changelog](src/CHANGES.md). -->

Use one of the following prefixes in the title of your pull request:

> -   `[ENH]` - enhancement of the specification that adds a new feature or

Reviewer: ```suggestion
-   `[ENH]` - enhancement of the software that adds a new feature or
```

    support for a new data type

> -   `[FIX]` - fix of a typo or language clarification

Reviewer: ```suggestion
-   `[FIX]` - fix of a bug or documentation error
```

> -   `[INFRA]` - changes to the infrastructure automating the specification

Reviewer: ```suggestion
-   `[INFRA]` - changes to the infrastructure automating the project
```

    release (for example, building HTML docs)
-   `[MISC]` - everything else including changes to the file listing
    contributors

If you are opening a pull request to obtain early feedback, but the changes are

> not ready to be merged (also known as a "work in progress" pull request, sometimes abbreviated by `WIP`)

Reviewer: ```suggestion
not ready to be merged (also known as a "work in progress" pull request, sometimes abbreviated by `WIP`),
```

please use a
[draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

A member of the BIDS-matlab team will review your changes to confirm that
they can be merged into the main codebase.

A [review](https://help.github.com/articles/about-pull-request-reviews/) will
usually consist of a few questions to help clarify the work you've done. 
Keep an eye on your GitHub notifications and be prepared to join in that
conversation.

> You can update your [fork](https://help.github.com/articles/about-forks/) of the

Reviewer: ```suggestion
You can update your [fork](https://help.github.com/articles/about-forks/) of
```

> BIDS Specification and the pull request will automatically update with those

Reviewer: ```suggestion
BIDS-matlab and the pull request will automatically update with those
```

commits. 
You don't need to submit a new pull request when you make a change in
response to a review.

GitHub has a [nice introduction](https://help.github.com/articles/github-flow/)
to the pull request workflow, but please [get in touch](#get-in-touch) if you
have any questions.

## Example pull request

<img align="center" src="https://i.imgur.com/s8yELfK.png" alt="Example-Contribution" width="800"/>

## Commenting on a pull request

> Our primary method of adding to or enhancing BIDS occurs in the form of

Reviewer: ```suggestion
Our primary method of adding to or enhancing BIDS-matlab occurs in the form of
```

This section outlines how to comment on a pull request.

### Navigating to open pull requests

The list of pull requests can be found by clicking on the "Pull requests" tab in
the
[BIDS-matlab repository](https://github.com/bids-standard/bids-matlab).

![BIDS-mainpage](commenting_images/BIDS_GitHub_mainpage.png "BIDS_GitHub_mainpage")

<!-- ### Selecting an open pull request

In this example we will be navigating to our
[BIDS common derivatives pull request](https://github.com/bids-standard/bids-specification/pull/265).

![BIDS-pr-list](commenting_images/BIDS_pr_list.png "BIDS_pr_list") -->

### Pull request description

Upon opening the pull request we see a detailed description of what this pull
request is seeking to address. 
Descriptions are important for reviewers and the community to gain context into 
what the pull request is achieving.

![BIDS-pr](commenting_images/BIDS_pr.png "BIDS_pr")

### Generally commenting on a pull request

At the bottom of the pull request page, a comment box is provided for general
comments and questions.

![BIDS-comment](commenting_images/BIDS_comment.png "BIDS-comment")

### Specific comments on a pull request

> The proposed changes to the text of the specification can be seen in the "Files

Reviewer: ```suggestion
The proposed changes to the software can be seen in the "Files
```

in the comment box (pictured below).

![BIDS-suggest-box](commenting_images/BIDS_suggest.png "BIDS-suggest")

Once the button is clicked the highlighted text will be copied into the comment
box and formatted as a
[Markdown code block](https://help.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks).

![BIDS-suggest-text](commenting_images/BIDS_suggest_text.png "BIDS-suggest-box")

The "Preview" tab in the comment box will show your suggestion as it will be
rendered. The "Suggested change" box will highlight the differences between the
original text and your suggestion.

![BIDS-suggest-change](commenting_images/BIDS_suggest_change.png "BIDS-suggest-change")

A comment may be submitted on its own by clicking "Add single comment". Several
comments may be grouped by clicking "Start a review". As more comments are
written, accept them with "Add review comment", and submit your review comments
as a batch by clicking the "Finish your review" button.

## Accepting suggestion from a review

When others are making [suggestions to your pull request](#suggesting-text), you
have the possibility to accept directly the changes suggested during the review

> through the github interface. This can often be faster and more convenient than

Reviewer: ```suggestion
through the GitHub interface. This can often be faster and more convenient than
```

You can add each suggestion one by one or group them together in a batch.

![BIDS_pr_accept_comment](commenting_images/BIDS_pr_accept_comment.png "BIDS_pr_accept_comment")

If you decide to batch the suggestions to add several of them at once, you must
scroll back to the top of the 'Files changed' page and the `commit suggestions`
button will let you add all those suggestions as a single commit.

![BIDS_pr_commit_batch](commenting_images/BIDS_pr_commit_batch.png "BIDS_pr_commit_batch")

Once those suggestions are committed the commit information should mention the
reviewer as a co-author.

![BIDS_pr_reviewer_credit](commenting_images/BIDS_pr_reviewer_credit.png "BIDS_pr_reviewer_credit")

## How the decision to merge a pull request is made?

The decision-making rules are outlined in
[DECISION-MAKING.md](DECISION-MAKING.md).

## Recognizing contributions

BIDS-matlab follows the
[all-contributors](https://github.com/kentcdodds/all-contributors)
specification, so we welcome and recognize all contributions from documentation
to testing to code development. 
You can see a list of current contributors in the
[README](./README.md).

> If you have made any type of contributions to the BIDS specifications, our team will add you as a contributor (or ask to be added if we forgot).

Reviewer: ```suggestion
If you have made any type of contributions to BIDS-matlab, our team will add you as a contributor (or ask to be added if we forgot).
```
